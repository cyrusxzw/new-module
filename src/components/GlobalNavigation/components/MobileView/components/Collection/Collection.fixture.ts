import type { CollectionProps } from './Collection.types';

const CollectionFixture: CollectionProps = {
  id: 'collection',
  label: 'Collection',
  title: 'Collection',
  items: [
    {
      id: 'ink',
      label: 'Link',
      title: 'Link',
      type: 'link' as const,
      url: '#link',
    },
  ],
};

export { CollectionFixture };
