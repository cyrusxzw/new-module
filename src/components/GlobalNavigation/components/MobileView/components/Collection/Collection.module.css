@import 'styles/index';

.backButton {
  justify-content: flex-start;
}

.directionIcon {
  flex: none;
  margin-right: 12px;

  &.dark {
    polygon {
      fill: var(--color-grey-45);
    }
  }

  &.light {
    polygon {
      fill: var(--color-grey-65);
    }
  }
}

.forward {
  margin-left: 12px;
  transform: rotate(-90deg) translateY(10px);
}

.backwards {
  margin-right: 12px;
  transform: rotate(90deg);
}
