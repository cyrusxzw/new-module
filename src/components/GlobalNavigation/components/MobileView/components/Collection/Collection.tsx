import React from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { useFocusOnFirst } from '~/customHooks';
import { Transition } from '~/components/Transition';
import { Button } from '~/components/Button';
import { Icon } from '~/components/Icon';
import { useGlobalNavigationStateContext } from '~/components/GlobalNavigation/GlobalNavigation.context';
import { useMobileViewContext } from '~/components/GlobalNavigation/components/MobileView/MobileView.context';
import { ListItem } from '~/components/GlobalNavigation/components/MobileView/components';
import type { CollectionType } from './Collection.types';
import compositionStyles from '../../MobileView.module.css';
import styles from './Collection.module.css';

const Collection: CollectionType = ({
  backLabel,
  id,
  isVisible,
  items,
  label,
  promotion,
  shouldHideTopCollection = false,
  title,
}) => {
  const {
    activeCollectionId,
    setActiveCollectionId,
  } = useGlobalNavigationStateContext();
  const { onBackButtonClick } = useMobileViewContext();
  const currentTheme = useThemeContext(null, 'dark');

  const isActive = activeCollectionId === id;
  const [listRef] = useFocusOnFirst(
    isActive,
    'a:not([tabindex="-1"]), button:not([tabindex="-1"])',
  );
  const handleOnClick = () => setActiveCollectionId(id);

  const handleOnBackButtonClick = () => {
    setActiveCollectionId('top');
    onBackButtonClick();
  };

  const listClassSet = cx(
    compositionStyles.list,
    compositionStyles.sublist,
    compositionStyles.slide,
  );

  const forwardIconClassSet = cx(
    styles.directionIcon,
    styles.forward,
    styles[currentTheme],
  );

  const backButtonClassSet = cx(
    compositionStyles.itemElement,
    styles.backButton,
  );

  const backIconClassSet = cx(
    styles.directionIcon,
    styles.backwards,
    styles[currentTheme],
  );

  return (
    <>
      {!shouldHideTopCollection && (
        <Button
          aria={{
            expanded: isActive,
            haspopup: true,
            hidden: !isActive,
          }}
          className={cx(compositionStyles.itemElement)}
          isInline={true}
          onClick={handleOnClick}
          tabIndex={activeCollectionId === 'top' && isVisible ? null : -1}
          theme={currentTheme}
          title={title}
        >
          {label}
          <Icon
            className={forwardIconClassSet}
            height={12}
            name="chevron"
            theme={currentTheme}
            width={12}
          />
        </Button>
      )}

      <Transition
        isActive={isActive && !!items && !!items.length}
        type="slideLeft"
      >
        <ul
          aria-hidden={!isActive}
          aria-label="submenu"
          className={listClassSet}
          ref={listRef}
        >
          <li className={compositionStyles.listItem}>
            <Button
              aria={{ hidden: !isActive, label: 'back' }}
              className={backButtonClassSet}
              isInline={true}
              onClick={handleOnBackButtonClick}
              tabIndex={isActive ? null : -1}
              theme={currentTheme}
              title={backLabel}
            >
              <Icon
                className={backIconClassSet}
                height={14}
                name="chevron"
                width={14}
              />{' '}
              {backLabel || label}
            </Button>
          </li>

          {items.map((props) => (
            <ListItem
              isVisible={isActive}
              itemProps={props}
              key={props.label}
            />
          ))}

          {promotion && <ListItem isVisible={isActive} itemProps={promotion} />}
        </ul>
      </Transition>
    </>
  );
};

export { Collection };
