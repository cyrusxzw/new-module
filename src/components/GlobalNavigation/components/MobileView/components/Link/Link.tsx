import React from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { Hyperlink } from '~/components/Hyperlink';
import { Icon } from '~/components/Icon';
import type { LinkType } from './Link.types';
import compositionStyles from '../../MobileView.module.css';
import styles from './Link.module.css';

const Link: LinkType = ({
  isTop,
  isNested,
  isVisible,
  label,
  title,
  url,
  isExternal,
}) => {
  const currentTheme = useThemeContext(null, 'dark');

  const classSet = cx(
    styles.base,
    styles[currentTheme],
    compositionStyles.itemElement,
    { [styles.top]: isTop },
    { [styles.nested]: isNested },
  );

  return (
    <Hyperlink
      className={classSet}
      tabIndex={!isVisible ? -1 : null}
      title={title}
      url={url}
    >
      {label}
      {isExternal && (
        <>
          {' '}
          <Icon height={14} name="rightArrow" theme={currentTheme} width={14} />
        </>
      )}
    </Hyperlink>
  );
};

export { Link };
