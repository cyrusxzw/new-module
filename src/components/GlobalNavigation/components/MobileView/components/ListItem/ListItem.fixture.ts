import type { ListItemProps } from './ListItem.types';

/* TODO{issue-10}: Improve fixture to be more reflective of tracking data
 * Currently implemented type to satisfy typescript for this fixture
 */

type NavTrackingPropsType = {
  menuType: 'Shop' | 'Read';
  panel: 'Panel 1' | 'Panel 2';
};

const navTrackingProps: NavTrackingPropsType = {
  menuType: 'Shop',
  panel: 'Panel 1',
};

const ListItemFixture: ListItemProps = {
  itemProps: {
    id: 'nested-collection',
    label: 'Nested Collection',
    title: 'Nested Collection',
    type: 'nested-collection' as const,
    items: [
      {
        id: 'nested-link',
        label: 'Nested Link',
        title: 'Nested Link',
        type: 'link' as const,
        url: '#nested-link',
        ...navTrackingProps,
      },
    ],
  },
  isVisible: true,
  isNestedItem: false,
};

export { ListItemFixture };
