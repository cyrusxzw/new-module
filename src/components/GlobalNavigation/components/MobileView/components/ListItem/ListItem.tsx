import React from 'react';
import cx from 'classnames';
import {
  Collection,
  Link,
  PromotionCard,
  NestedCollection,
} from '../../components';
import type { ListItemType } from './ListItem.types';
import compositionStyles from '../../MobileView.module.css';

const ListItem: ListItemType = ({
  isNestedItem,
  isTopItem = false,
  isVisible = false,
  itemProps,
}) => {
  let returnElement = null;

  if (itemProps.type === 'collection') {
    returnElement = <Collection {...{ ...itemProps, isVisible }} />;
  } else if (itemProps.type === 'nested-collection') {
    returnElement = <NestedCollection {...{ ...itemProps, isVisible }} />;
  } else if (itemProps.type === 'promotion-card') {
    returnElement = <PromotionCard {...{ ...itemProps, isVisible }} />;
  } else if (itemProps.type === 'link') {
    returnElement = (
      <Link {...{ ...itemProps, isVisible, isNestedItem, isTopItem }} />
    );
  } else {
    return null;
  }

  const classSet = cx(compositionStyles.listItem, {
    [compositionStyles.nestedItem]: itemProps.type === 'nested-collection',
    [compositionStyles.nestedCollectionLink]:
      itemProps.type === 'link' && isNestedItem,
  });

  return <li className={classSet}>{returnElement}</li>;
};

export { ListItem };
