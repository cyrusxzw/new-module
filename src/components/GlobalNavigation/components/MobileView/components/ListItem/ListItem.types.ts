import type { ComponentWithChildren } from '~/types';
import type {
  Promotion,
  Collection,
  Link,
  Read,
  NestedCollection,
  NotableNestedCollection,
} from '~/components/GlobalNavigation/GlobalNavigation.types';

type ListItemProps = {
  isNestedItem?: boolean;
  isTopItem?: boolean;
  isVisible?: boolean;
  itemProps:
    | NestedCollection
    | Collection
    | Link
    | Promotion
    | NotableNestedCollection
    | Read;
};

type ListItemType = ComponentWithChildren<ListItemProps>;

export type { ListItemType };
