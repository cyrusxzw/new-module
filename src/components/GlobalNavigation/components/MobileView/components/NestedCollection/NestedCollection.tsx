import React from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { Button } from '~/components/Button';
import { useGlobalNavigationStateContext } from '~/components/GlobalNavigation/GlobalNavigation.context';
import { useMobileViewContext } from '~/components/GlobalNavigation/components/MobileView/MobileView.context';
import { useVariableHeightStyle } from '~/components/GlobalNavigation/components/MobileView/MobileView.hooks';
import { ListItem } from '~/components/GlobalNavigation/components/MobileView/components';
import type { NestedCollectionType } from './NestedCollection.types';
import compositionStyles from '../../MobileView.module.css';
import styles from './NestedCollection.module.css';

const NestedCollection: NestedCollectionType = ({
  id,
  isVisible,
  items = [],
  label,
  title,
}) => {
  const { activeCollectionId } = useGlobalNavigationStateContext();
  const {
    activeNestedCollectionIds,
    onNestedCollectionClick,
  } = useMobileViewContext();
  const currentTheme = useThemeContext(null, 'dark');
  const isActive = activeNestedCollectionIds.includes(id);
  const { ref, style } = useVariableHeightStyle(isActive);
  const handleOnNestedClick = () => onNestedCollectionClick(id);

  const listClassSet = cx(
    compositionStyles.list,
    compositionStyles.sublist,
    styles.collapse,
  );

  return (
    <>
      <Button
        aria={{
          expanded: isActive,
          haspopup: true,
          hidden: !isActive,
        }}
        className={cx(compositionStyles.itemElement, styles.nestedButton)}
        isInline={true}
        onClick={handleOnNestedClick}
        tabIndex={activeCollectionId === 'top' || !isVisible ? -1 : null}
        theme={currentTheme}
        title={title}
      >
        {label}
        <span
          className={cx(
            styles.icon,
            { [styles.expanded]: isActive },
            styles[currentTheme],
          )}
        />
      </Button>

      <ul
        aria-hidden={!isActive}
        aria-label="submenu"
        className={listClassSet}
        ref={ref}
        style={style}
      >
        {items.map((props) => (
          <ListItem
            isNestedItem={true}
            isVisible={isActive}
            itemProps={props}
            key={props.label}
          />
        ))}
      </ul>
    </>
  );
};

export { NestedCollection };
