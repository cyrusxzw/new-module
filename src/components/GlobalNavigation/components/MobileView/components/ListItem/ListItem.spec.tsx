import React from 'react';
import { render } from '@testing-library/react';
import { ListItem } from './ListItem';
import { ListItemFixture } from './ListItem.fixture';

import * as collectionFile from '../Collection/Collection';
import * as nestedCollectionFile from '../NestedCollection/NestedCollection';
import * as promotionCardFile from '../../../PromotionCard/PromotionCard';
import * as linkFile from '../Link/Link';

describe('<GlobalNavigation.MobileView.ListItem />', () => {
  const collectionSpy: jest.SpyInstance = jest
    .spyOn(collectionFile, 'Collection')
    .mockImplementation(() => <div>collection</div>);
  const nestedCollectionSpy: jest.SpyInstance = jest
    .spyOn(nestedCollectionFile, 'NestedCollection')
    .mockImplementation(() => <div>nested collection</div>);
  const promotionCardSpy: jest.SpyInstance = jest
    .spyOn(promotionCardFile, 'PromotionCard')
    .mockImplementation(() => <div>promotion</div>);
  const linkSpy: jest.SpyInstance = jest
    .spyOn(linkFile, 'Link')
    .mockImplementation(() => <div>link</div>);

  it('should render base component correctly as nested collection', () => {
    const { container } = render(<ListItem {...ListItemFixture} />);

    expect(container).toMatchSnapshot();
  });

  it('should render collection', () => {
    const collectionListItemFixture = { ...ListItemFixture };
    collectionListItemFixture.itemProps.type = 'collection';
    const { container } = render(<ListItem {...collectionListItemFixture} />);

    expect(container).toMatchSnapshot();
  });

  it('should render notable collection', () => {
    const notableCollectionListItemFixture = { ...ListItemFixture };
    notableCollectionListItemFixture.itemProps.type =
      'notable-nested-collection';
    const { container } = render(
      <ListItem {...notableCollectionListItemFixture} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render promotion card', () => {
    const promoCardListItemFixture = { ...ListItemFixture };
    promoCardListItemFixture.itemProps.type = 'promotion';
    const { container } = render(<ListItem {...promoCardListItemFixture} />);

    expect(container).toMatchSnapshot();
  });

  it('should render link', () => {
    const linkListItemFixture = { ...ListItemFixture };
    linkListItemFixture.itemProps.type = 'link';
    const { container } = render(<ListItem {...linkListItemFixture} />);

    expect(container).toMatchSnapshot();
  });
});
