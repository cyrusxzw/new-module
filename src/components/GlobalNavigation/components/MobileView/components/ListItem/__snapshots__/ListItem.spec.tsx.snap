// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalNavigation.MobileView.ListItem /> should render base component correctly as nested collection 1`] = `
<div>
  <li
    class="listItem nestedItem"
  >
    <div>
      nested collection
    </div>
  </li>
</div>
`;

exports[`<GlobalNavigation.MobileView.ListItem /> should render collection 1`] = `
<div>
  <li
    class="listItem"
  >
    <div>
      collection
    </div>
  </li>
</div>
`;

exports[`<GlobalNavigation.MobileView.ListItem /> should render link 1`] = `
<div>
  <li
    class="listItem"
  >
    <div>
      link
    </div>
  </li>
</div>
`;

exports[`<GlobalNavigation.MobileView.ListItem /> should render notable collection 1`] = `
<div>
  <li
    class="listItem nestedItem"
  >
    <div>
      nested collection
    </div>
  </li>
</div>
`;

exports[`<GlobalNavigation.MobileView.ListItem /> should render promotion card 1`] = `
<div>
  <li
    class="listItem"
  >
    <div>
      promotion
    </div>
  </li>
</div>
`;
